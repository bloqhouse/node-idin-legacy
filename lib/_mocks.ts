const constantDate = '2017-10-09T20:08:32.110Z';

(Date as any) = class extends Date {
  constructor() {
    super(constantDate)
  }
}
