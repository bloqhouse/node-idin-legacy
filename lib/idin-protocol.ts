import { FileKeyInfo, KeyInfoProvider, SignedXml, xpath } from 'xml-crypto'
import { DOMParser } from 'xmldom'
import { PRIVATE_KEY, PUBLIC_KEY_FINGERPRINT } from './constants'

export function signXML(xml: string, selfVerify: boolean = false) {
  const sig = new SignedXml()
  const entryPoint = '/*'
  const transformers = ['http://www.w3.org/2000/09/xmldsig#enveloped-signature', 'http://www.w3.org/2001/10/xml-exc-c14n#']
  const xmlenc = 'http://www.w3.org/2001/04/xmlenc#sha256'
  const signatureAlgorithm = 'http://www.w3.org/2001/04/xmldsig-more#rsa-sha256'
  const MyKeyInfo = function(this: any, key: string) {
    this.getKeyInfo = (privateKey: any, prefix: string) => {
      return `<KeyName>${PUBLIC_KEY_FINGERPRINT}</KeyName>`
    }
  } as any

  sig.addReference(entryPoint, transformers, xmlenc, void 0, void 0, void 0, true)
  sig.keyInfoProvider = new MyKeyInfo(PRIVATE_KEY)
  sig.signatureAlgorithm = signatureAlgorithm
  sig.signingKey = PRIVATE_KEY
  sig.computeSignature(xml)
  const res = sig.getSignedXml()
  selfVerify ? verifyOwnSignature(res, xml) : Function.prototype()
  return res
}

export function verifyOwnSignature(res: string, xml: string) {
  const doc = new DOMParser().parseFromString(res)
  const signature = xpath(doc, '//*[local-name(.)=\'Signature\' and namespace-uri(.)=\'http://www.w3.org/2000/09/xmldsig#\']')[0]
  const sx = new SignedXml()
  sx.keyInfoProvider = new FileKeyInfo('cert.pem')
  sx.loadSignature(signature)
  const result = sx.checkSignature(xml)
  /* istanbul ignore next */
  if (!result) {
    throw new Error(sx.validationErrors)
  }
}
