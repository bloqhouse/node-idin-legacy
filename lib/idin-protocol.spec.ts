import './_mocks'
import { signXML } from './idin-protocol'

const xml = '<?xml version="1.0" encoding="UTF-8"?><x></x>'

describe('format idin protocol', () => {
  it('it works correctly', () => {
    expect(signXML(xml)).toMatchSnapshot()
    expect(signXML(xml, true)).toMatchSnapshot()
  })
})
