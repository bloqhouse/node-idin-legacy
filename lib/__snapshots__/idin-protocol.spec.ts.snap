// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format idin protocol it works correctly 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><x><Signature xmlns=\\"http://www.w3.org/2000/09/xmldsig#\\"><SignedInfo><CanonicalizationMethod Algorithm=\\"http://www.w3.org/2001/10/xml-exc-c14n#\\"/><SignatureMethod Algorithm=\\"http://www.w3.org/2001/04/xmldsig-more#rsa-sha256\\"/><Reference URI=\\"\\"><Transforms><Transform Algorithm=\\"http://www.w3.org/2000/09/xmldsig#enveloped-signature\\"/><Transform Algorithm=\\"http://www.w3.org/2001/10/xml-exc-c14n#\\"/></Transforms><DigestMethod Algorithm=\\"http://www.w3.org/2001/04/xmlenc#sha256\\"/><DigestValue>9hy1oK7rXCJu4rTqLZ7cGUH3rPyGm4QllC8VRv6mX60=</DigestValue></Reference></SignedInfo><SignatureValue>XIxFrdB1chFmnVt8/umPcLFgUnojCfIe5Er4McVq7fuo0K8jMsg7vKZ+ZRusaALHn8FnEVREm5xemfkvfiQLmLV5iqCESZbKrdxTnfLRsB3JLDKBb3q+BYNbL1RoHOhPa8FxeMtmPMJWd5HnNGXyxkprjr7ZHkZ4KG9MuW0klNbJEjEnL38ikcUHepI5SPOym6Tk4r+VDOwgieX0ynEMF/7bRas5yH5kCk2q7mBQV6xL2apQ+JYO3TsZeu+VAHK43lgBF02iPnO5NE7sCym9CX/DMb9U4pljgwVWvU6Dps5RaWPayib1cOXX8LjULuClsikG3IqETgqb0csKMInotQ==</SignatureValue><KeyInfo><KeyName>8c6415e72e553d2a9bda7afc9fcc0a5ec5d82841</KeyName></KeyInfo></Signature></x>"`;

exports[`format idin protocol it works correctly 2`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><x><Signature xmlns=\\"http://www.w3.org/2000/09/xmldsig#\\"><SignedInfo><CanonicalizationMethod Algorithm=\\"http://www.w3.org/2001/10/xml-exc-c14n#\\"/><SignatureMethod Algorithm=\\"http://www.w3.org/2001/04/xmldsig-more#rsa-sha256\\"/><Reference URI=\\"\\"><Transforms><Transform Algorithm=\\"http://www.w3.org/2000/09/xmldsig#enveloped-signature\\"/><Transform Algorithm=\\"http://www.w3.org/2001/10/xml-exc-c14n#\\"/></Transforms><DigestMethod Algorithm=\\"http://www.w3.org/2001/04/xmlenc#sha256\\"/><DigestValue>9hy1oK7rXCJu4rTqLZ7cGUH3rPyGm4QllC8VRv6mX60=</DigestValue></Reference></SignedInfo><SignatureValue>XIxFrdB1chFmnVt8/umPcLFgUnojCfIe5Er4McVq7fuo0K8jMsg7vKZ+ZRusaALHn8FnEVREm5xemfkvfiQLmLV5iqCESZbKrdxTnfLRsB3JLDKBb3q+BYNbL1RoHOhPa8FxeMtmPMJWd5HnNGXyxkprjr7ZHkZ4KG9MuW0klNbJEjEnL38ikcUHepI5SPOym6Tk4r+VDOwgieX0ynEMF/7bRas5yH5kCk2q7mBQV6xL2apQ+JYO3TsZeu+VAHK43lgBF02iPnO5NE7sCym9CX/DMb9U4pljgwVWvU6Dps5RaWPayib1cOXX8LjULuClsikG3IqETgqb0csKMInotQ==</SignatureValue><KeyInfo><KeyName>8c6415e72e553d2a9bda7afc9fcc0a5ec5d82841</KeyName></KeyInfo></Signature></x>"`;
